src/analyses/variable-sensitivity/abstract_object.cpp

```diff
@@ -0,0 +1,299 @@
+  out_any_modifications=m->merge_state(abstract_object_pointert(this), op);
+  return m;
+}
+
+/*******************************************************************\
+
+Function: abstract_objectt::expression_transform
+
+  Inputs:
+   expr - the expression to evaluate and find the result of it. this will
+          be the symbol referred to be op0()
+
+ Outputs: Returns the abstract_object representing the result of this expression
+          to the maximum precision available.
+
+ Purpose: To try and resolve different expressions with the maximum level
+          of precision available.
+
+\*******************************************************************/
+
+abstract_object_pointert abstract_objectt::expression_transform(
+  const exprt &expr,
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  exprt constant_replaced_expr=expr;
+  constant_replaced_expr.operands().clear();
+  for(const exprt &op : expr.operands())
+  {
+    abstract_object_pointert lhs_abstract_object=environment.eval(op, ns);
+    const exprt &lhs_value=lhs_abstract_object->to_constant();
+
+    if(lhs_value.id()==ID_nil)
+    {
+      // One of the values is true so we can't really do anything more with
+      // this expression and should just return top for the result
+      return abstract_object_pointert(
+        new abstract_objectt(expr.type(), true, false));
+    }
+    else
+    {
+      // rebuild the operands list with constant versions of
+      // any symbols
+      constant_replaced_expr.operands().push_back(lhs_value);
+    }
+  }
+
+  exprt simplified=simplify_expr(constant_replaced_expr, ns);
+  if(simplified.is_constant())
+  {
+    constant_exprt constant_expr=to_constant_expr(simplified);
+
+    // TODO(tkiley): This should be going through the abstract_object_factory
+    // but at the moment this produces a two value abstraction for type bool
+    // so for now we force it to be the constant abstraction
+    return abstract_object_pointert(
+      new constant_abstract_valuet(constant_expr));
+  }
+  else
+  {
+    return environment.abstract_object_factory(expr.type(), true, false);
+  }
+}
+
+/*******************************************************************\
+
+Function: abstract_objectt::is_top
+
+  Inputs:
+
+ Outputs: Returns true if the abstract object is representing the top (i.e. we
+          don't know anything about the value).
+
+ Purpose: Find out if the abstract object is top
+
+\*******************************************************************/
+
+bool abstract_objectt::is_top() const
+{
+  return top;
+}
+
+/*******************************************************************\
+
+Function: abstract_objectt::is_bottom
+
+  Inputs:
+
+ Outputs: Returns true if the abstract object is representing the bottom.
+
+ Purpose: Find out if the abstract object is bottom
+
+\*******************************************************************/
+
+bool abstract_objectt::is_bottom() const
```

Comment:
Elsewhere there is `get_is_bottom` (and `get_is_top`) - consistent naming would be appreciated.